sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java

```diff
@@ -0,0 +1,364 @@
+          schema
+              .columns()
+              .stream()
+              .filter(x -> !x.materializedOrAlias())
+              .map(x -> quoteIdentifier(x.name()))
+              .collect(Collectors.joining(", "));
+      return "INSERT INTO " + quoteIdentifier(table) + " (" + columnsStr + ")";
+    }
+
+    @Setup
+    public void setup() throws SQLException {
+      connection = new ClickHouseDataSource(jdbcUrl(), properties()).getConnection();
+      executor =
+          Executors.newSingleThreadExecutor(
+              new ThreadFactoryBuilder().setNameFormat("clickhouse-jdbc-%d").build());
+      queue = new ArrayBlockingQueue<>(QUEUE_SIZE);
+      bundleFinished = new AtomicBoolean(false);
+    }
+
+    @StartBundle
+    public void startBundle() throws SQLException {
+      statement = connection.createStatement();
+      bundleFinished.set(false);
+
+      // When bundle starts, we open statement and stream data in bundle.
+      // When we finish bundle, we close statement.
+
+      // For streaming, we create a background thread for http client,
+      // and we feed it through BlockingQueue.
+
+      insert =
+          executor.submit(
+              () -> {
+                try {
+                  statement.sendRowBinaryStream(
+                      insertSql(schema(), table()),
+                      stream -> {
+                        while (true) {
+                          Row row;
+                          try {
+                            row = queue.poll(1, TimeUnit.SECONDS);
+                          } catch (InterruptedException e) {
+                            break; // time to go
+                          }
+
+                          if (row == null) {
+                            if (bundleFinished.get() && queue.isEmpty()) {
+                              // bundle is finished and queue is drained, time to go
+                              break;
+                            } else {
+                              // timeout, let's try again
+                            }
+                          } else {
+                            ClickHouseWriter.writeRow(stream, schema(), row);
+                          }
+                        }
+                      });
+                  statement.close();
+                } catch (Exception e) {
+                  throw new RuntimeException(e);
+                } finally {
+                  try {
+                    if (!statement.isClosed()) {
+                      // happens only if statement wasn't closed in the end of the previous block
+                      statement.close();
+                    }
+                  } catch (SQLException e) {
+                    LOG.error("Failed to close statement", e);
+                  }
+                }
+              });
+    }
+
+    @FinishBundle
+    public void finishBundle() throws ExecutionException, InterruptedException {
+      bundleFinished.set(true);
+      insert.get(); // drain queue
```

Comment:
Deduplication is done on the server side. When a block is inserted, replicas put checksum of the block to a ring buffer on top of ZooKeeper. In our case, one bundle is one block, so deduplication should work as soon as bundles don't change between retries, because even splitting a bundle will break deduplication. I see there is `@RequiresStableInput` but wasn't sure if it is needed in this case.